File: drum-game-web/src/playfield/notation/RenderGroup.ts
// group of notes that can be rendered to a canvas
// meant to be grouped each ~4 beats (but not necessarily by measure)

import { Note } from "../../interfaces/BJson"
import Beatmap from "../../utils/Beatmap"
import ChannelInfo from "../../utils/ChannelInfo";
import NotationDisplay from "./NotationDisplay";

const stemHeight = 2.5;
const AugmentationDotGap = 0.1;
const AccentGap = 0.1;

// contains multiple NoteGroups
export default class RenderGroup {
    NoteGroups: NoteGroup[] = []

    Geometry: [] | undefined = undefined;

    ComputeGeometry() { // basically prepares for Render() ahead of time
        this.Geometry = []
    }

    Index: number;
    constructor(index: number) {
        this.Index = index;
    }

    first = true;

    Render(display: NotationDisplay, context: CanvasRenderingContext2D) {
        if (!this.Geometry) this.ComputeGeometry();
        const font = display.Font;
        const engravingDefaults = font.engravingDefaults;
        const spacing = display.Spacing;
        for (const group of this.NoteGroups) {
            const flags = group.flags;
            const down = group.voice === 1;
            const dir = down ? 1 : -1;
            const beamHeight = -dir * engravingDefaults.beamThickness

            const targetHeight = group.highestNote + dir * stemHeight

            let beamLeft = Number.MAX_VALUE;
            let beamRight = Number.MIN_VALUE;

            // there's like 200 lines of code we need to add here
            for (const flag of flags) {
                let bottomNote = flag.notes[0];
                for (const note of flag.notes) {
                    if (down ? note.noteMapping[0] < bottomNote.noteMapping[0] : note.noteMapping[0] > bottomNote.noteMapping[0])
                        bottomNote = note;
                }
                const bottomAnchor = display.GetNoteheadAnchor(bottomNote.noteMapping[1], down);
                const bottomX = flag.beat * spacing; // this is where we will render the bottom note for certain

                let accent = false;

                for (const note of flag.notes) {
                    accent ||= note.modifier === "accent"
                    const l = note.noteMapping;
                    const anchor = display.GetNoteheadAnchor(l[1], down)[0];
                    const noteX = bottomX + bottomAnchor[0] - anchor;
                    context.fillText(l[1], noteX, l[0])
                    if (note.modifier === "ghost")
                        context.fillText("\uE0CE", noteX, l[0])

                    // handle note dotting
                    if (flag.duration == 0.75 || flag.duration == 0.375) {
                        const rightSide = down ? font.glyphsWithAnchors[ChannelInfo.CodepointMap[l[1]]].stemUpSE[0] : anchor;
                        let dotY = l[0];
                        console.log(dotY)
                        if (Number.isInteger(dotY)) dotY += dir / 2; // prevent dot being placed on a line
                        context.fillText("\uE1E7", noteX + rightSide + AugmentationDotGap, l[0])
                    }
                }
                let flagLeft = bottomX + bottomAnchor[0];
                flag.flagLeft = flagLeft;
                if (!down) flagLeft -= engravingDefaults.stemThickness;
                beamLeft = Math.min(beamLeft, flagLeft);
                beamRight = Math.max(beamRight, flagLeft + engravingDefaults.stemThickness);
                // stem
                context.fillRect(flagLeft, bottomNote.noteMapping[0] - bottomAnchor[1], engravingDefaults.stemThickness,
                    targetHeight - bottomNote.noteMapping[0] + bottomAnchor[1]);


                let flagGlyph: any = undefined;
                if (flags.length == 1) // fancy flag
                {
                    if (flag.duration! <= 0.375)
                        flagGlyph = down ? "\uE243" : "\uE242";
                    else if (flag.duration! <= 0.75)
                        flagGlyph = down ? "\uE241" : "\uE240";
                    if (flagGlyph) {
                        // this extension should technically be applied to targetHeight prior to this
                        const extension = down ? 0.5 : -0.5;
                        const y = targetHeight + extension;
                        const name = ChannelInfo.CodepointMap[flagGlyph];
                        const anchor = down ? font.glyphsWithAnchors[name].stemDownSW : font.glyphsWithAnchors[name].stemUpNW;
                        context.fillText(flagGlyph, flagLeft - anchor[0], y + anchor[1]);
                    }
                }
                if (accent) {
                    const accentHeight = targetHeight + dir * (AccentGap +
                        (flags.length > 1 || flagGlyph ? engravingDefaults.beamThickness : 0));
                    context.fillText(down ? "\uE4A1" : "\uE4A0", bottomX, accentHeight);
                }
            }

            if (flags.length > 1) { // beam
                context.fillRect(beamLeft, targetHeight, beamRight - beamLeft, -beamHeight);
                let beamY = targetHeight - dir * (engravingDefaults.beamSpacing);
                let duration = 0.25;
                // depth 1 = 16th, depth 2 = 32nd
                for (let depth = 1; depth <= 2; depth++) {
                    let added = false;
                    let beamStart = 0;
                    let beamCount = 0;
                    for (let i = 0; i < flags.length; i++) {
                        const flag = flags[i];
                        if (flag.duration! <= duration || flag.duration! == duration * 1.5) {
                            if (beamCount == 0) beamStart = flag.flagLeft;
                            beamCount += 1;
                        } else {
                            // there's a note that is not part of the current beam, so we have to end it
                            if (beamCount > 0) {
                                var end = beamCount == 1 ? (beamStart + flag.flagLeft + engravingDefaults.stemThickness) / 2 : flags[i - 1].flagLeft;

                                context.fillRect(beamStart, beamY, end - beamStart, beamHeight);
                                added = true;
                                beamCount = 0;
                            }
                        }
                    }
                    // we finished going through the notes and there is an active beam
                    if (beamCount > 0) {
                        if (beamCount == 1) {
                            const a = beamStart - spacing * duration * 0.5
                            context.fillRect(a, beamY, beamStart - a, beamHeight);
                            added = true;
                        }
                        else {
                            var end = flags[flags.length - 1].flagLeft;
                            context.fillRect(beamStart, beamY, beamStart - end, beamHeight);
                            added = true;
                        }
                    }
                    if (!added) break; // if we didn't add any beams, there won't be any more
                    duration *= 0.5;
                    beamY += dir * -(engravingDefaults.beamSpacing + engravingDefaults.beamThickness);
                }
            }
        }
        this.first = false;
    }

    // start and end are the RenderGroup index, with index 0 = beats 0-4 = ticks 0-57600
    // note, only the NoteGroup's tick is used for assigning the render group index
    // since a NoteGroup can be up to 1 full beat, it could start a 3.999 and push a full beat into the next group's rendering area
    // see `ReloadNotes` in MusicNotationBeatmapDisplay.cs for how to calculate the index from a tick
    static BuildRenderGroups(beatmap: Beatmap, startIndex = 0, endIndex?: number) {
        const o: RenderGroup[] = []

        let currentGroup = (undefined as any) as RenderGroup;
        let currentNoteGroups: { [key: number]: NoteGroup } = {}
        let currentBeat = Number.MIN_VALUE;

        for (const note of beatmap.HitObjects) {
            if (note.time < 0) continue;

            // TODO this will not work after we add MeasureChanges, see `ReloadNotes`
            // with fractional measures, this can be something like 3.5
            const beat = Math.floor(note.time);

            const container = Math.floor(beat / 4);
            while (container !== (o.length - 1)) { // push current/empty containers until we get to where we need to be
                currentGroup = new RenderGroup(o.length);
                o.push(currentGroup);
            }

            if (beat !== currentBeat) {
                currentNoteGroups = {};
                currentBeat = beat;
            }

            const voice = note.voice;
            if (!(voice in currentNoteGroups)) {
                // TODO end is not correct with MeasureChanges
                const newGroup: NoteGroup = { flags: [], beat, voice, highestNote: note.noteMapping[0], end: beat + 1 };
                currentGroup.NoteGroups.push(newGroup);
                currentNoteGroups[voice] = newGroup
            }
            addNote(currentNoteGroups[voice], note)
        }

        for (const e of o) {
            for (const group of e.NoteGroups) {
                for (let i = 0; i < group.flags.length; i++) {
                    if (i === group.flags.length - 1) {
                        group.flags[i].duration = group.end - group.flags[i].beat
                    } else {
                        group.flags[i].duration = group.flags[i + 1].beat - group.flags[i].beat
                    }
                }
            }
        }

        return o
    }
}

function addNote(noteGroup: NoteGroup, note: Note) {
    const lastFlag = noteGroup.flags[noteGroup.flags.length - 1];
    if (note.voice === 1) {
        if (note.noteMapping[0] > noteGroup.highestNote)
            noteGroup.highestNote = note.noteMapping[0]
    } else {
        if (note.noteMapping[0] < noteGroup.highestNote)
            noteGroup.highestNote = note.noteMapping[0]
    }
    if (!lastFlag || lastFlag.beat !== note.time) {
        noteGroup.flags.push({ beat: note.time, notes: [note] } as Flag)
    } else {
        lastFlag.notes.push(note);
    }
}

interface NoteGroup {
    flags: Flag[]
    highestNote: number
    beat: number // this number is what is used to push NoteGroups into RenderGroups. Simply divide it by 4 and round down to get the RenderGroup
    end: number // used to calculate flag durations
    voice: number
}
interface Flag {
    notes: Note[]
    beat: number
    duration?: number
    flagLeft: number
}

File: DrumGame/DrumGame-Game/Notation/RenderGroup.cs
using System;
using DrumGame.Game.Beatmaps;
using DrumGame.Game.Utils;
using osu.Framework.Extensions.EnumExtensions;
using osu.Framework.Graphics;
using osu.Framework.Graphics.Containers;
using osu.Framework.Graphics.Shapes;
using osu.Framework.Graphics.Sprites;

namespace DrumGame.Game.Notation;

public partial class MusicFont
{
    public void RenderGroup(Container container, NoteGroup group, double groupOffset, Container outside)
    {
        var down = group.Down;
        var dir = down ? 1 : -1;
        var flags = group.Flags;

        // this is the height where we will draw the top bar of the group
        var targetHeight = group.HighestNote / 2f + (down ? StemHeight : -StemHeight);
        // left-most flag position, used for drawing top beam
        var beamLeft = float.MaxValue;
        // right-most flag position, used for drawing top beam
        var beamRight = float.MinValue;

        for (var i = 0; i < flags.Count; i++)
        {
            // Compute final flag values before rendering
            var flag = flags[i];
            if (i == flags.Count - 1)
            {
                flag.EffectiveDuration = group.GroupEnd - flags[i].Time;
            }
            else
            {
                flag.EffectiveDuration = flags[i + 1].Time - flags[i].Time;
            }

            // anchor offset for the bottom notehead of this flag

            // we use the X value for the standard notehead for consistency
            // this is to prevent the flag from shifting when different noteheads appears as the bottom note
            // this is really bad for the ride-bell notehead
            // good example: bbbbb (sixteenth notes)
            //               s s s
            // where b is ride bell, s is snare - the spacing on the ride notes will be awful
            var bottomAnchor = (GetNoteheadAnchor(MusicGlyph.noteheadBlack, down).x,
                GetNoteheadAnchor(flag.BottomNote.Note.Notehead, down).y);

            // this is the X position of the bottom note
            // all other notes in this flag group will be positioned based on this
            var bottomX = (float)((flag.Time - groupOffset) * Spacing);

            var bottomNote = flag.BottomNote;
            flag.FlagLeft = bottomX + bottomAnchor.x;
            if (!down) flag.FlagLeft -= EngravingDefaults.stemThickness;

            foreach (var note in flag.Notes)
            {
                var skinNote = note.Note;

                // we don't care about the y-anchor since that is only used for the bottom note drawing the stem
                var anchor = GetNoteheadAnchor(skinNote.Notehead, down).x;

                // left-most part of the notehead
                var noteX = bottomX + bottomAnchor.x - anchor;
                var l = note.Modifiers.HasFlag(NoteModifiers.Left);
                var r = note.Modifiers.HasFlag(NoteModifiers.Right);
                var ghost = note.Modifiers.HasFlag(NoteModifiers.Ghost);

                var noteheadColor = skinNote.Color;
                if (skinNote.AccentColor != default && note.Modifiers.HasFlag(NoteModifiers.Accented))
                    noteheadColor = skinNote.AccentColor;
                else if (skinNote.GhostColor != default && ghost)
                    noteheadColor = skinNote.GhostColor;

                if (skinNote.StickingColorNotehead)
                {
                    if (l)
                        noteheadColor = skinNote.LeftColor;
                    else if (r)
                        noteheadColor = skinNote.RightColor;
                }

                container.Add(Notehead(skinNote, noteX, noteheadColor));
                if ((l || r) && !skinNote.StickingColorNotehead)
                {
                    var hollow = skinNote.IsHollow();
                    container.Add(new SpriteText
                    {
                        Text = l ? "L" : "R",
                        Origin = Anchor.Centre,
                        Colour = l ? skinNote.LeftColor : skinNote.RightColor,
                        Depth = -skinNote.Position - 5,
                        Scale = new osuTK.Vector2(hollow ? 2f / 20 : 1f / 20),
                        Y = skinNote.Position / 2f,
                        X = noteX + 0.5f
                    });
                }
                if (note.Duration > 0) // roll
                {
                    var centerOffset = 0.5f;
                    outside.Add(new Circle
                    {
                        Y = skinNote.Position / 2f - RollHeight / 2f,
                        // this is placed outside, so we can ignore groupOffset
                        X = (float)(flag.Time * Spacing) + bottomAnchor.x - anchor + centerOffset,
                        Height = RollHeight,
                        Width = (float)(note.Duration * Spacing) - centerOffset,
                        Colour = RollColour,
                        Depth = 10
                    });
                }
                if (ghost)
                {
                    container.Add(new NoteSprite
                    {
                        Character = MusicGlyph.noteheadParenthesis.Codepoint(),
                        Y = -8 + skinNote.Position / 2f,
                        X = noteX,
                        Colour = Colour,
                        Font = FontUsage
                    });
                }
                // need to dot note
                // if ((BitConverter.DoubleToInt64Bits(flag.EffectiveDuration) & 0xfffffffffffffL) == 0x8000000000000L)
                if (flag.EffectiveDuration == 0.75 || flag.EffectiveDuration == 0.375)
                {
                    // right-most part of the notehead
                    var rightSide = down ? GetAnchorValue(skinNote.Notehead, "stemUpSE")[0] : anchor;
                    var dotY = skinNote.Position;
                    if ((dotY & 1) == 0) dotY += dir; // prevent dot being placed on a line
                    container.Add(new NoteSprite
                    {
                        Character = MusicGlyph.augmentationDot.Codepoint(),
                        Y = -8 + dotY / 2f,
                        X = noteX + rightSide + AugmentationDotGap,
                        Colour = Colour,
                        Font = FontUsage
                    });
                }
            }
            var flagX = flag.FlagLeft;
            beamLeft = Math.Min(beamLeft, flagX);
            beamRight = Math.Max(beamRight, flagX + EngravingDefaults.stemThickness);
            container.Add(new Box // stem
            {
                X = flagX,
                Y = bottomNote.Note.Position / 2f - bottomAnchor.y,
                Origin = down ? Anchor.TopLeft : Anchor.BottomLeft,
                Colour = Colour,
                Width = EngravingDefaults.stemThickness,
                Height = Math.Abs(bottomNote.Note.Position / 2f - targetHeight) + dir * bottomAnchor.y
            });
            Drawable flagGlyph = null;
            if (flags.Count == 1) // fancy flag
            {
                flagGlyph = RenderFlag(flags[0], targetHeight, down);
                if (flagGlyph != null) container.Add(flagGlyph);
            }
            if (flag.Accented)
            {
                var accentHeight = targetHeight + dir * (AccentGap +
                    (flags.Count > 1 || flagGlyph != null ? EngravingDefaults.beamThickness : 0));
                container.Add(new NoteSprite
                {
                    Character = (down ? MusicGlyph.articAccentBelow : MusicGlyph.articAccentAbove).Codepoint(),
                    Y = accentHeight - 8,
                    X = bottomX,
                    Colour = Colour,
                    Font = FontUsage
                });
            }
        }

        if (flags.Count > 1)
        {
            container.Add(Beam(beamLeft, beamRight, down ? targetHeight : targetHeight - EngravingDefaults.beamThickness));
            // depth 1 = 16th, depth 2 = 32nd
            var beamY = targetHeight + (down ? -EngravingDefaults.beamSpacing - EngravingDefaults.beamThickness : EngravingDefaults.beamSpacing);
            var duration = 0.25f;
            foreach (var depth in new[] { 1, 2 })
            {
                var startCount = container.Count;
                float beamStart = 0;
                var beamCount = 0;
                for (var i = 0; i < flags.Count; i++)
                {
                    var flag = flags[i];
                    if (flag.EffectiveDuration <= duration || flag.EffectiveDuration == duration * 1.5f)
                    {
                        if (beamCount == 0) beamStart = flag.FlagLeft;
                        beamCount += 1;
                    }
                    else
                    {
                        // there's a note that is not part of the current beam, so we have to end it
                        if (beamCount > 0)
                        {
                            var end = beamCount == 1 ? (beamStart + flag.FlagLeft + EngravingDefaults.stemThickness) / 2f : flags[i - 1].FlagLeft;
                            container.Add(Beam(beamStart, end, beamY));
                            beamCount = 0;
                        }
                    }
                }
                // we finished going through the notes and there is an active beam
                if (beamCount > 0)
                {
                    if (beamCount == 1)
                    {
                        container.Add(Beam(beamStart - Spacing * duration * 0.5f, beamStart, beamY));
                    }
                    else
                    {
                        var end = flags[flags.Count - 1].FlagLeft;
                        container.Add(Beam(beamStart, end, beamY));
                    }
                }
                if (container.Count == startCount) break; // if we didn't add any beams, there won't be any more
                duration *= 0.5f;
                beamY += dir * -(EngravingDefaults.beamSpacing + EngravingDefaults.beamThickness);
            }
        }
    }
}

